### ISSUE TITLE
🤖 Mammalian enhancer dataset via halLiftover on the Zoonomia 241-way Cactus

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/bolinas-dna#121]
## Goal

Build a **dataset of enhancers in mammals** by projecting high-quality human enhancer annotations onto other Zoonomia mammals using the 241-way Cactus whole-genome alignment (`halLiftover`). This is an alternate path to #96 — same downstream objective (cross-species enhancer labels to feed into gLM training), different route: **evolutionary alignment instead of a learned classifier.**

## Relationship to #96

| Aspect | #96 (classifier path) | This issue (halLiftover path) |
|---|---|---|
| Signal source | Sequence features learned from human/mouse ENCODE cCREs | Whole-genome alignment (Zoonomia Cactus) |
| Per-species enhancer call | Classifier scores each window in the target genome | halLiftover projects each human enhancer onto the target |
| Species coverage | Any genome with a FASTA | Only the 241 Zoonomia mammals |
| Clade-specific enhancers (e.g. rodent-specific) | May be recovered if the classifier generalizes | **Missed** — only enhancers present in the source species (human) are projected. Individual species-specific enhancers are negligible; clade-specific ones are the real blind spot. Mitigable by adding mouse ENCODE SCREEN as a second source (optional follow-up). |
| Lineage-specific loss of function | Correct by construction | False positive (projects coordinates even where function is gone) |
| Compute cost | Training + genome-wide inference | One-off halLiftover pass |
| Extensibility to non-mammals / non-Zoonomia species | Yes | No |

Both paths produce the same kind of output (per-species enhancer BEDs) and plug into the training-dataset pipeline. Running both lets us compare training performance and test whether alignment-based or classifier-based labels are the cleaner supervision signal for gLM pretraining.

## Assumptions and rationale

1. **Human enhancer annotations are the gold standard.** ENCODE SCREEN cCREs (dELS, pELS) filtered with primate/mammalian conservation (phastCons / phyloP) and repeat/functional-element masks — same positive-set recipe as discussed in #96 — are the highest-confidence input we have.
2. **Cactus alignments preserve orthology for the regulatory regions we care about.** The Zoonomia 241-way Cactus is the state-of-the-art mammalian WGA; halLiftover on this HAL is the documented Weng-lab procedure for projecting hg38 cCREs onto the other 240 species ([Andrews et al. 2023 Supp. Methods](https://www.science.org/doi/10.1126/science.abn7930)). Conserved enhancers are exactly the ones Cactus will align well — the method's precision is self-selecting for the class we want.
3. **Precision > recall** (same as #96): a missed lineage-specific enhancer is cheaper than a false-positive non-functional sequence polluting pretraining.

## Prior work — what exists, what doesn't

From reading Christmas et al. 2023, Andrews et al. 2023, and Moore et al. 2026 (main + supp):

- **Per-species cCRE BED files are NOT publicly released.** Andrews et al. Data S1 is a single hg38 spreadsheet annotated with per-cCRE N1/N2 counts. The per-species lifted BEDs are mentioned in their Methods ("saved the results in BED format") but **not among the supplementary releases.** Their code to regenerate them is on Zenodo ([10.5281/zenodo.7447627](https://doi.org/10.5281/zenodo.7447627)).
- **Zoonomia HAL and 2bit genomes are public:** `http://cgl.gi.ucsc.edu/data/cactus/241-mammalian-2020v2-hub/hub.txt` (hundreds of GB to ~1 TB).
- **Assembly list** is in Christmas et al. Supp Table S1 (mostly `.1`-era GCA/GCF accessions — *different* from the current NCBI RefSeq assemblies used in the [main training pipeline](snakemake/training_dataset/dataset_creation/config/config.yaml)).
- **No Zoonomia-assembly ↔ current-RefSeq liftover (chain file or equivalent) exists anywhere public, and cross-assembly mismatch is not discussed in any of the three papers.**

## Design decision: use Zoonomia assemblies directly, as a parallel pipeline

Rather than building within-species chains to project Zoonomia coordinates onto current NCBI RefSeq assemblies, we **operate entirely in the Zoonomia assembly space**:

- Extract per-species FASTAs from the Zoonomia HAL with `hal2fasta`.
- Run halLiftover on the human enhancer BED to produce per-species BEDs on the Zoonomia assembly of each target species.
- Build the training dataset directly from those Zoonomia-assembly FASTAs + BEDs.

This is an **experimental parallel pipeline** in its own Snakemake directory. We deliberately accept code duplication with the main training-dataset pipeline — no refactor into shared library code yet. If the halLiftover approach proves useful we can consolidate later; if not, we delete this directory cleanly.

Consequence: this pipeline does not use the NCBI RefSeq GTF / RepeatMasker annotations that the main pipeline relies on for CDS / UTR / promoter / ncRNA masks. If we need comparable functional-region masks here, we'll skip them or take them from the Zoonomia hub where available.

## Species scope

Start with **one representative species per placental order**, ensuring **Homo sapiens (Primates)** and **Mus musculus (Rodentia)** are included. This gives ~20 species with broad phylogenetic coverage, enough to see whether the approach produces usable training data, without committing to all 240 upfront. Expanded species lists are a follow-up.

## Proposed pipeline

New parallel pipeline `snakemake/mammalian_enhancers_halliftover/`:

```
config/
  config.yaml                   # species list (one per order), HAL URL, source enhancer BED filters
workflow/
  Snakefile
  envs/                         # cactus-binaries (halLiftover, hal2fasta), ucsc-tools, bedtools
  rules/
    download.smk                # Zoonomia HAL, hg38 reference, source enhancer inputs
    build_source_enhancers.smk  # hg38 ENCODE SCREEN ELS + conservation/repeat/CDS filters
    extract_fastas.smk          # hal2fasta per target species
    halliftover.smk             # hg38 enhancers → each Zoonomia target assembly
    qc.smk                      # coverage stats, lost-interval breakdown, per-species summaries
    package.smk                 # per-species BED + FASTA in a format the training pipeline can consume
```

## Key design questions (for discussion)

1. **Source enhancer set.** Same filtered ENCODE SCREEN ELS + conservation masking as the #96 positive set, for apples-to-apples comparison? Or also explore broader sets (e.g. unfiltered dELS/pELS) to understand how the filter interacts with alignability?
2. **halLiftover parameters.** Minimum fraction of the source interval that must align? Handling of split mappings? Andrews et al. describe their parameters in their supp methods — worth reusing for comparability.
3. **Validation.** Sanity-check by halLiftover'ing *within* hg38 (expect ~100% recovery) and by projecting coding exons (should recover near-perfectly) before trusting enhancer coordinates.
4. **HAL storage.** The HAL is hundreds of GB to ~1 TB. Do we pull it once to a shared location, or stream / use UCSC public halLiftover endpoints if available?

## Tracking

This issue is the living document for this analysis. Description = current state; comments = append-only iteration log. (Same convention as #120.)

## References

- **Zoonomia 241-way Cactus** — Christmas et al. (2023). *Science* 380, eabn3943. [10.1126/science.abn3943](https://doi.org/10.1126/science.abn3943).
- **halLiftover-on-Cactus for cCREs (methodology)** — Andrews et al. (2023). *Science* 380, eabn7930. [10.1126/science.abn7930](https://doi.org/10.1126/science.abn7930). Code: [Zenodo 10.5281/zenodo.7447627](https://doi.org/10.5281/zenodo.7447627).
- **ENCODE V4 cCREs** — Moore et al. (2026). *Nature*. [10.1038/s41586-025-09909-9](https://doi.org/10.1038/s41586-025-09909-9).
- **Cactus** — Armstrong et al. (2020). *Nature* 587, 246–251. [10.1038/s41586-020-2871-y](https://doi.org/10.1038/s41586-020-2871-y).
- **halLiftover / HAL format / hal2fasta** — Hickey et al. (2013). *Bioinformatics* 29, 1341–1342. [10.1093/bioinformatics/btt128](https://doi.org/10.1093/bioinformatics/btt128).

**Comment by gonzalobenegas** [comment]
Closed by https://github.com/Open-Athena/bolinas-dna/pull/158
